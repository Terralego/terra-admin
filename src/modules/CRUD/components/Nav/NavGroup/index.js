import NavGroup from './NavGroup';

export default NavGroup;
