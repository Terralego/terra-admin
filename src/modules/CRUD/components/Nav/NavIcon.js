import React from 'react';

const NavIcon = ({ src, alt = '', ...rest }) => src && (
  <span className="bp3-icon" {...rest}>
    <img className="CRUD-nav__icon" src={src} alt={alt} />
  </span>
);

export default NavIcon;
