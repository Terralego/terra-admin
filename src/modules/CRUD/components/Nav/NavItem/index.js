import { connectAuthProvider } from '@terralego/core/modules/Auth';
import compose from '../../../../../utils/compose';

import NavItem from './NavItem';

const authProviderGetter = ({
  authenticated,
  user,
}) => {
  const permissions = authenticated ? user.permissions : [];
  return {
    displayAddFeature: permissions.includes('can_add_feature'),
  };
};

export default compose(
  connectAuthProvider(authProviderGetter),
)(NavItem);
