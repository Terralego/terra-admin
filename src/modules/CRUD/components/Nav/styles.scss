.CRUD-nav {
  display: flex;
  flex-direction: column;
  width: 6rem;
  max-height: calc(100vh - 50px);
  
  &--active {
    width: 17rem;
    .CRUD-nav__button {
      text-align: right;
      border: 0;
      padding-right: 1.5rem;
    }
  }
  
  &__button {
    padding: 0.7rem 1rem 0;
    text-align: center;
  }
  
  &__menu {
    &#{&} {
      flex: auto;
      position: relative;
      margin-top: 0.5rem;
      padding: 0.5rem 1rem 1rem;
      overflow: auto;
      border-top: 1px solid #efefef;
    }
  }

  &__group,
  &__item-content {
    .bp3-icon {
      width: 20px;
      margin-right: 7px;
    }
  }

  &__group-button {
    width: 100%;
  }

  &__popover {
    left:-20px;
  }

  &__icon {
    max-width: 100%;
  }

  &__group:not(:first-child) {
    margin-top: 0.5rem;
    padding-top: 0.5rem;
    border-top: 1px solid #efefef;
  }

  &__list {
    margin-left: 1rem !important;
    border-left: 1px solid #efefef;
    overflow: hidden;
  }

  &__item {
    padding: 0.25rem 0.3rem;
  }

  &__action {
    display: flex;
  }
  
  &__link {
    position: relative;
    display: inline-block;
    margin-right: auto;
    &:hover,
    &:focus {
      outline: none;
      position: relative;
      z-index: 100;
      .bp3-button {
        background: #e4e9ed !important;
      }
      .bp3-button-text {
        opacity: 1;
      }
    }
  }
}
