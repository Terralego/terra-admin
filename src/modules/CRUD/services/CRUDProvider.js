import React from 'react';
import connect from 'react-ctx-connect';
import { fetchSettings } from './CRUD';
import {
  fetchFeaturesList,
  fetchFeature as fetchFeatureAction,
  saveFeature as saveFeatureAction,
  deleteFeature as deleteFeatureAction,
  updateFeatureIdentifier,
  fetchCustomEndpoint,
} from './features';

import { saveAttachmentCategories } from './attachments';

export const CRUDContext = React.createContext({});
export const connectCRUDProvider = connect(CRUDContext);

const { Provider } = CRUDContext;

const PAGE_SIZE = 999;

export class CRUDProvider extends React.Component {
  state = {
    settings: undefined,
    featuresList: {},
    feature: {},
    mapConfig: undefined,
    errors: {
      attachmentCategories: undefined,
      settings: undefined,
      featuresList: [],
      feature: [],
    },
  };

  dataTableRef = React.createRef();

  detailsRef = React.createRef();

  componentWillUnmount () {
    this.isUnmount = true;
  }

  setMap = map => !this.isUnmount && this.setState({ map });

  getSettings = async endpoint => {
    const result = {};
    try {
      const settings = await fetchSettings(endpoint);
      result.settings = settings;
    } catch (e) {
      result.error = e;
    }

    const { settings = {}, error } = result;

    this.setState(({ errors }) => ({
      settings,
      errors: {
        ...errors,
        settings: error,
      },
    }));

    return settings;
  }

  getFormattedError = ({
    store,
    error,
    ids,
  }) => {
    const { errors: { [store]: errorStore } } = this.state;

    const propsFromIds = Object.keys(ids);

    // Remove error if it was before in the store
    if (!(error instanceof Error)) {
      return errorStore.filter(errorItem => (
        propsFromIds.every(prop => errorItem[prop] !== ids[prop])
      ));
    }

    // Modify error if it was before in the store
    const isErrorItemAlreadyExisting = errorStore.some(errorItem => (
      propsFromIds.every(prop => errorItem[prop] === ids[prop])
    ));

    if (isErrorItemAlreadyExisting) {
      return errorStore.map(errorItem => (
        propsFromIds.every(prop => errorItem[prop] === ids[prop])
          ? { error, ...ids }
          : errorItem
      ));
    }

    // Else add error in the store
    return [...errorStore, { error, ...ids }];
  }

  getFeaturesList = async (layerId, querystring) => {
    const result = {};
    try {
      const featuresList = await fetchFeaturesList(layerId, querystring);
      result.featuresList = featuresList;
    } catch (e) {
      result.error = e;
    }
    const { featuresList = [], error = {} } = result;

    this.setState(({ errors }) => ({
      featuresList,
      errors: {
        ...errors,
        featuresList: this.getFormattedError({ error, ids: { layerId }, store: 'featuresList' }),
      },
    }));
    return featuresList.length && featuresList;
  };

  fetchFeature = async (layerId, featureId) => {
    const result = {};
    try {
      const feature = await fetchFeatureAction(layerId, featureId);
      result.feature = feature;
    } catch (e) {
      result.error = e;
    }

    const { feature: newFeature = {}, error = {} } = result;

    this.setState(({ errors, feature }) => ({
      feature: updateFeatureIdentifier(feature, newFeature),
      errors: {
        ...errors,
        feature: this.getFormattedError({ error, ids: { layerId, featureId }, store: 'feature' }),
      },
    }));

    return Object.keys(newFeature).length > 0 && newFeature;
  }

  saveFeature = async (layerId, featureId, data, method = 'PUT') => {
    const result = {};
    try {
      const feature = await saveFeatureAction(layerId, featureId, data, method);
      result.feature = feature;
    } catch (e) {
      result.error = e;
    }

    const { feature: newFeature = {}, error = {} } = result;

    this.setState(({ errors, feature }) => ({
      feature: updateFeatureIdentifier(feature, newFeature),
      errors: {
        ...errors,
        feature: this.getFormattedError({ error, ids: { layerId, featureId }, store: 'feature' }),
      },
    }));

    return result;
  }

  deleteFeature = async (layerId, featureId) => {
    const result = {};
    try {
      await deleteFeatureAction(layerId, featureId);
      result.feature = featureId;
    } catch (e) {
      // Error
    }
    const { feature } = result;

    this.setState(({ featuresList, feature: { [feature]: deletedFeature, ...rest } }) => ({
      feature: rest,
      featuresList: {
        ...featuresList,
        results: deletedFeature
          ? featuresList.results.filter(({ identifier }) => identifier !== featureId)
          : featuresList.results,
      },
    }));

    return feature;
  }

  getAttachmentCategories = async endpoint => {
    const { attachmentCategories } = this.state;
    if (attachmentCategories) {
      return attachmentCategories;
    }
    const result = {};
    try {
      const categories = await fetchCustomEndpoint(`${endpoint}`, {
        querystring: { page_size: PAGE_SIZE },
      });
      result.attachmentCategories = categories;
    } catch (e) {
      result.error = e;
    }

    const { attachmentCategories: nextAttachmentCategories, error } = result;

    this.setState(({ errors }) => ({
      attachmentCategories: nextAttachmentCategories,
      errors: {
        ...errors,
        attachmentCategories: error,
      },
    }));

    return nextAttachmentCategories;
  }

  createAttachmentCategories = async name => {
    const result = {};
    try {
      const category = await saveAttachmentCategories(name);
      result.category = category;
    } catch (e) {
      result.error = e;
    }

    const { category: newCategory = {}, error = {} } = result;

    this.setState(({ errors, attachmentCategories }) => ({
      attachmentCategories: {
        ...attachmentCategories,
        newCategory,
      },
      errors: {
        ...errors,
        attachmentCategories: {
          ...attachmentCategories.error,
          error,
        },
      },
    }));

    return Object.keys(newCategory).length > 0 && newCategory;
  }

  findOrCreateAttachmentCategory = async category => {
    const { name, id } = category;
    if (id !== null) {
      return category;
    }
    const {
      attachmentCategories = [],
      settings: {
        config: {
          attachment_categories: endpoint,
        },
      },
    } = this.state;
    let categories = attachmentCategories;
    if (!categories.length) {
      const { results } = await this.getAttachmentCategories(endpoint);
      categories = results;
    }
    const categoryInList = categories.find(cat => (
        cat?.name.toLowerCase() === name.trim().toLowerCase()
    ));
    if (categoryInList !== undefined) {
      return categoryInList;
    }
    const newCategory = await this.createAttachmentCategories(name);
    return newCategory;
  }

  render () {
    const { children } = this.props;

    const {
      dataTableRef,
      detailsRef,
      getMapConfig,
      getSettings,
      getFeaturesList,
      fetchFeature,
      saveFeature,
      deleteFeature,
      getAttachmentCategories,
      findOrCreateAttachmentCategory,
      setMap,
    } = this;
    const value = {
      ...this.state,
      dataTableRef,
      detailsRef,
      getMapConfig,
      getSettings,
      getFeaturesList,
      fetchFeature,
      saveFeature,
      deleteFeature,
      getAttachmentCategories,
      findOrCreateAttachmentCategory,
      setMap,
    };

    return (
      <Provider value={value}>
        {children}
      </Provider>
    );
  }
}

export default CRUDProvider;
