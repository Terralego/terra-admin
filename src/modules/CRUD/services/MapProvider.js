import React, { useState, useCallback, useRef, useEffect } from 'react';
import connect from 'react-ctx-connect';
import { useTranslation } from 'react-i18next';

import {
  DEFAULT_CONTROLS,
  CONTROL_CAPTURE,
  CONTROL_BACKGROUND_STYLES,
  CONTROL_SEARCH,
  CONTROLS_TOP_RIGHT,
} from '@terralego/core/modules/Map';

import { getBounds } from './features';

const BASE_SEARCH_API_PROVIDER = 'https://nominatim.openstreetmap.org/search?';

const CONTROL_LIST = [{
  control: CONTROL_BACKGROUND_STYLES,
  position: CONTROLS_TOP_RIGHT,
}, {
  control: CONTROL_CAPTURE,
  position: CONTROLS_TOP_RIGHT,
}];

const sortByOrder = ({ order: a = 0 }, { order: b = 0 }) => a - b;

export const MapContext = React.createContext({});
export const connectMapProvider = connect(MapContext);

const { Provider } = MapContext;

export const MapProvider = ({ children }) => {
  const detailsRef = useRef(null);
  const dataTableRef = useRef(null);

  const [map, setMap] = useState(null);

  const { i18n } = useTranslation();

  const setFitBounds = useCallback(({ coordinates, hasDetails }) => {
    if (!coordinates.length || !map) return;

    const { current: detail } = detailsRef;
    const { current: dataTable } = dataTableRef;

    const padding = {
      top: 20,
      right: hasDetails ? (detail?.offsetWidth ?? 0 + 50) : 50,
      bottom: !hasDetails ? (dataTable?.offsetHeight ?? 0 + 20) : 20,
      left: 20,
    };
    map.resize();
    map.fitBounds(getBounds(coordinates), { padding, duration: 0 });
  }, [dataTableRef, detailsRef, map]);

  const onSearch = useCallback(async query => {
    const params = new URLSearchParams();
    params.append('q', query);
    params.append('format', 'geojson');
    params.append('accept-language', i18n.language);

    const headers = new Headers([['Content-Type', 'application/json']]);
    let results;
    try {
      results = await fetch(`${BASE_SEARCH_API_PROVIDER}${params}`, {
        headers,
      }).then(response => response.json());
    } catch (e) {
      return null;
    }
    // Filter to avoid duplicates location
    const filteredFeatures = results.features.reduce(
      (list, result) => {
        if (!list.some(item => item.properties.display_name === result.properties.display_name)) {
          list.push(result);
        }
        return list;
      }, [],
    );
    const data = [{
      total: filteredFeatures.length,
      results: filteredFeatures.map(
        ({ bbox, properties: { osm_id: id, display_name: label } }) => ({
          label,
          id,
          bounds: bbox,
        }),
      ),
    }];
    return data;
  }, [i18n.language]);

  const onSearchResultClick = useCallback(({ result: { bounds } }) => {
    const [xMin, yMin, xMax, yMax] = bounds;
    setFitBounds({
      coordinates: [[xMin, yMin], [xMax, yMax]],
      hasDetails: Boolean(detailsRef.current),
    });
  }, [setFitBounds]);

  const [controls, setControls] = useState([...DEFAULT_CONTROLS, ...CONTROL_LIST]);
  const [interactiveMapProps, setInteractiveMapProps] = useState(undefined);
  const [layers, setLayers] = useState([]);

  useEffect(() => {
    if (!map) {
      return;
    }
    setControls(prevControls => {
      if (prevControls.some(({ control }) => control === CONTROL_SEARCH)) {
        return prevControls;
      }
      return [{
        control: CONTROL_SEARCH,
        position: CONTROLS_TOP_RIGHT,
        maxResults: 10,
        onSearch,
        onSearchResultClick,
        order: -1,
      }, ...prevControls].sort(sortByOrder);
    });
  }, [map, onSearch, onSearchResultClick]);

  const addControl = useCallback(controlToAdd => {
    if (!controlToAdd) return;
    setControls(prevControls => {
      if (prevControls.some(({ control }) => control === controlToAdd.control)) {
        return prevControls;
      }
      return [controlToAdd, ...prevControls].sort(sortByOrder);
    });
  }, []);

  const removeControl = useCallback(controlToRemove => {
    if (!controlToRemove) return;
    setControls(prevControls => {
      if (prevControls.some(({ control }) => control === controlToRemove)) {
        return prevControls.filter(({ control }) => control !== controlToRemove);
      }
      return prevControls;
    });
  }, []);

  const featureToHighlight = useCallback(({ featureId, layer, hover }) => {
    const { id: layerId, source } = layers.find(({ 'source-layer': sourceLayer }) => sourceLayer === layer) || {};
    if (!map || !featureId || !layerId) {
      return;
    }
    const { addHighlight, removeHighlight } = interactiveMapProps;
    if (hover) {
      addHighlight && addHighlight({
        layerId,
        featureId,
        highlightColor: 'red',
        unique: true,
        source,
      });
    } else {
      removeHighlight && removeHighlight({
        layerId,
        featureId,
      });
    }
  }, [interactiveMapProps, layers, map]);


  const value = {
    addControl,
    controls,
    detailsRef,
    dataTableRef,
    featureToHighlight,
    layers,
    interactiveMapProps,
    map,
    setControls,
    setFitBounds,
    setMap,
    setLayers,
    setInteractiveMapProps,
    removeControl,
  };

  return (
    <Provider value={value}>
      {children}
    </Provider>
  );
};

export default MapProvider;
