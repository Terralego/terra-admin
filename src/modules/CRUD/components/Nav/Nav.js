import React, { useState, useEffect, useMemo, useContext } from 'react';
import { Button } from '@blueprintjs/core';
import { useTranslation } from 'react-i18next';
import classnames from 'classnames';

import { CRUDContext } from '../../services/CRUDProvider';
import { MapContext } from '../../services/MapProvider';
import NavGroup from './NavGroup';

import './styles.scss';

const sortBy = prop => ({ [prop]: orderA }, { [prop]: orderB }) =>
  (typeof orderA === 'number' ? orderA : Infinity) - (typeof orderB === 'number' ? orderB : Infinity);

const sortByOrder = sortBy('order');

export const Nav = () => {
  const { map } = useContext(MapContext);
  const { settings: { menu } = {} } = useContext(CRUDContext);

  const { t } = useTranslation();

  const [menuOpen, setMenuOpen] = useState(true);

  const orderedMenu = useMemo(() => (
    menu && menu
      .reduce((group, menuItem) => {
        const { id, crud_views: views } = menuItem;
        return [
          ...group,
          ({ views: views.sort(sortByOrder), collapsed: (id !== null), ...menuItem }),
        ];
      }, [])
      .filter(({ crud_views: views }) => views.length)
      .sort(sortByOrder)
  ), [menu]);

  useEffect(() => {
    map && map.resize();
  }, [map, menuOpen]);

  if (!menu) {
    return null;
  }

  return (
    <div
      className={classnames(
        'CRUD-nav',
        { 'CRUD-nav--active': menuOpen },
      )}
    >
      <div
        className="CRUD-nav__button"
      >
        <Button
          icon={menuOpen ? 'menu-closed' : 'menu-open'}
          aria-controls="CRUD-nav__menu"
          expandable={menuOpen.toString()}
          aria-label={t(menuOpen ? 'CRUD.nav.foldMenu' : 'CRUD.nav.unfoldMenu')}
          onClick={() => setMenuOpen(prevMenOpen => !prevMenOpen)}
          minimal
        />
      </div>
      {orderedMenu && (
        <ul
          id="CRUD-nav__menu"
          className="CRUD-nav__menu"
        >
          {orderedMenu.map(groupList => (
            <NavGroup
              key={groupList.id}
              navCollapsed={!menuOpen}
              {...groupList}
            />
          ))}
        </ul>
      )}
    </div>
  );
};

export default Nav;
