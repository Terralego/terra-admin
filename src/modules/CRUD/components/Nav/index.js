import { withTranslation } from 'react-i18next';

import { connectCRUDProvider } from '../../services/CRUDProvider';
import Nav from './Nav';


export default connectCRUDProvider('settings')(
  withTranslation()(Nav),
);
