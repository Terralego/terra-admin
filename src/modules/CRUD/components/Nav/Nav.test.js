import React from 'react';
import renderer from 'react-test-renderer';
import { MapContext } from '../../services/MapProvider';
import { CRUDContext } from '../../services/CRUDProvider';

import Nav from './Nav';

jest.mock('@blueprintjs/core', () => ({
  Button:  ({ children, ...rest }) => <div {...rest}>{children}</div>,
}));

jest.mock('./NavGroup', () => props => <div {...props} />);

jest.mock('@terralego/core/modules/Map', () => ({}));

const settings = {
  menu: [{
    id: 1,
    crud_views: [{
      id: 5,
      name: 'Monuments',
      order: 0,
    }, {
      id: 1,
      name: 'Sentiers',
      order: 1,
    }, {
      id: 3,
      name: 'Zones tampons',
      pictogram: null,
      order: 3,
    }],
    name: 'Biens UNESCO',
    order: 0,
  }, {
    id: 8,
    crud_views: [{
      id: 23,
      name: 'Chemins',
      pictogram: null,
      order: 0,
    }],
    name: 'Autres',
    order: 1,
    pictogram: null,
  }, {
    id: null,
    name: 'Unclassified',
    order: null,
    pictogram: null,
    crud_views: [{
      id: 2,
      name: 'Biens',
      pictogram: null,
      order: 1,
    }, {
      id: 4,
      name: 'Biens polygone',
      pictogram: null,
      order: 4,
    }],
  }],
};

it('should snapshot correctly', () => {
  const tree = renderer.create((
    <MapContext.Provider value={{ map: jest.fn() }}>
      <CRUDContext.Provider value={{ settings }}>
        <Nav />
      </CRUDContext.Provider>
    </MapContext.Provider>
  )).toJSON();
  expect(tree).toMatchSnapshot();
});
