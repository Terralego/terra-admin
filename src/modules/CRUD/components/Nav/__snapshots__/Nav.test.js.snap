// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should snapshot correctly 1`] = `
<div
  className="CRUD-nav CRUD-nav--active"
>
  <div
    className="CRUD-nav__button"
  >
    <div
      aria-controls="CRUD-nav__menu"
      aria-label="CRUD.nav.foldMenu"
      expandable="true"
      icon="menu-closed"
      minimal={true}
      onClick={[Function]}
    />
  </div>
  <ul
    className="CRUD-nav__menu"
    id="CRUD-nav__menu"
  >
    <div
      collapsed={true}
      crud_views={
        Array [
          Object {
            "id": 5,
            "name": "Monuments",
            "order": 0,
          },
          Object {
            "id": 1,
            "name": "Sentiers",
            "order": 1,
          },
          Object {
            "id": 3,
            "name": "Zones tampons",
            "order": 3,
            "pictogram": null,
          },
        ]
      }
      id={1}
      name="Biens UNESCO"
      navCollapsed={false}
      order={0}
      views={
        Array [
          Object {
            "id": 5,
            "name": "Monuments",
            "order": 0,
          },
          Object {
            "id": 1,
            "name": "Sentiers",
            "order": 1,
          },
          Object {
            "id": 3,
            "name": "Zones tampons",
            "order": 3,
            "pictogram": null,
          },
        ]
      }
    />
    <div
      collapsed={true}
      crud_views={
        Array [
          Object {
            "id": 23,
            "name": "Chemins",
            "order": 0,
            "pictogram": null,
          },
        ]
      }
      id={8}
      name="Autres"
      navCollapsed={false}
      order={1}
      pictogram={null}
      views={
        Array [
          Object {
            "id": 23,
            "name": "Chemins",
            "order": 0,
            "pictogram": null,
          },
        ]
      }
    />
    <div
      collapsed={false}
      crud_views={
        Array [
          Object {
            "id": 2,
            "name": "Biens",
            "order": 1,
            "pictogram": null,
          },
          Object {
            "id": 4,
            "name": "Biens polygone",
            "order": 4,
            "pictogram": null,
          },
        ]
      }
      id={null}
      name="Unclassified"
      navCollapsed={false}
      order={null}
      pictogram={null}
      views={
        Array [
          Object {
            "id": 2,
            "name": "Biens",
            "order": 1,
            "pictogram": null,
          },
          Object {
            "id": 4,
            "name": "Biens polygone",
            "order": 4,
            "pictogram": null,
          },
        ]
      }
    />
  </ul>
</div>
`;
